src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs

```diff
@@ -1125,4 +1127,172 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
+            if (marshallerType == null)
+            {
+                // TypeLoadException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            if (marshallerType.IsGenericDefinition)
+            {
+                // TypeLoadException throw here.
+                ThrowHelper.ThrowTypeLoadException(marshallerType);
+            }
+
+            var customMarshallerType = Context.SystemModule.GetKnownType("System.Runtime.InteropServices", "ICustomMarshaler");
+            if (customMarshallerType == null)
+            {
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            if (!marshallerType.CanCastTo(customMarshallerType))
+            {
+                // ApplicationException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            var getInstanceMethod = marshallerType.GetMethod(
+                "GetInstance",
+                new MethodSignature(MethodSignatureFlags.Static, 0, customMarshallerType, new[] { Context.GetWellKnownType(WellKnownType.String) }));
+            if (getInstanceMethod == null)
+            {
+                // ApplicationException throw here.
+                ThrowHelper.ThrowMarshalDirectiveException();
+            }
+
+            var getInstanceMethodType = Context.GetFunctionPointerType(
+                new MethodSignature(MethodSignatureFlags.Static, 0, Context.GetWellKnownType(WellKnownType.Object),
+                    new TypeDesc[] { Context.GetWellKnownType(WellKnownType.String) }));
+            var initializeCustomMarshallerMethod = InteropTypes.GetPInvokeMarshal(Context).GetKnownMethod("InitializeCustomMarshaller",
+                new MethodSignature(MethodSignatureFlags.Static, 0, Context.GetWellKnownType(WellKnownType.Object),
+                    new TypeDesc[] {
+                        Context.GetWellKnownType(WellKnownType.RuntimeTypeHandle),
+                        Context.GetWellKnownType(WellKnownType.RuntimeTypeHandle),
+                        Context.GetWellKnownType(WellKnownType.String),
+                        getInstanceMethodType,
+                    }
+                ));
+
+            ILEmitter emitter = _ilCodeStreams.Emitter;
+            var lMarshaller = emitter.NewLocal(customMarshallerType);
+            var cookie = MarshalAsDescriptor.Cookie;
+
+            // Custom marshaller initialization should not be catched, so initialize early
+            ILCodeStream fnptrLoadStream = _ilCodeStreams.FunctionPointerLoadStream;
+            fnptrLoadStream.Emit(ILOpcode.ldtoken, emitter.NewToken(ManagedParameterType));
+            fnptrLoadStream.Emit(ILOpcode.ldtoken, emitter.NewToken(marshallerType));
+            fnptrLoadStream.Emit(ILOpcode.ldstr, emitter.NewToken(cookie));
+            fnptrLoadStream.Emit(ILOpcode.ldftn, emitter.NewToken(getInstanceMethod));
```

Comment:
If you drop the `if (getInstanceMethod == null)` check a couple lines above, you can do this:

```suggestion
            if (getInstanceMethod != null)
            {
                fnptrLoadStream.Emit(ILOpcode.ldftn, emitter.NewToken(getInstanceMethod));
            }
            else
            {
                fnptrLoadStream.EmitLdc(0);
                fnptrLoadStream.Emit(ILOpcode.conv_i);
            }
```

Then this will go into the `throw new ApplicationException();` codepath you have in the classlib.